File: CodingSamples/DotNet/Fundamentals/Functional/DelegateTest/DemoApp/Investment.cs
delegate double InterestRate(int period);

class Investment
{
    //a read-only property which can only be set in instance initializer
    public double Installment { get; init; } = 50000;

    //a property defined with required keyword must be set in instance initializer
    public required int Years { get; init; }

    public double FutureValue(InterestRate scheme)
    {
        double i = scheme.Invoke(Years) / 100;
        return (Installment / i) * (Math.Pow(1 + i, Years) - 1);
    }
}

File: CodingSamples/Java/Fundamentals/Functional/MethodReferenceTest/demo/Investment.java
class Investment {

    private double installment;
    private int years;

    public Investment(double installment, int years) {
        this.installment = installment;
        this.years = years;
    }

    public double futureValue(InterestRate rate) {
        double i = rate.forPeriod(years) / 100;
        return (installment / i) * (Math.pow(1 + i, years) - 1);
    }
}

